### DISCUSSION TITLE
I am getting a + symbol inserted instead of a space

### DISCUSSION THREAD

**Discussion by JediThumper** [SchneiderInfosystems/FB4D#175] [Q&A] [answered]
Hi Christoph,

I don't know why, But since upgrading to 1.6 in getting a plus symbol instead of a space, in the document name  - not fields.

This has changed since I update my code to use the full path in doc rather than the older deprecated style.

My current code:

procedure TFBAccess.SaveLastTimeUsed(Atime: TDateTime);
var
  Doc: IFirestoreDocument;
begin
  Doc := TFirestoreDocument.Create(['Last Used', LocalSettingsObject.Email], ProjectKey);
  Doc.AddOrUpdateField(TJSONObject.SetTimeStamp('Last_Used', Atime));
  fDataBase.InsertOrUpdateDocument(Doc, nil, OnSaveLastTimeUsedInsert, OnSaveLastTimeUsedrError)
end;

<img width="533" alt="Screenshot 2024-04-27 at 17-37-49 Denneroll - Cloud Firestore - Data - Firebase console" src="https://github.com/SchneiderInfosystems/FB4D/assets/71242335/685cb84e-2eb7-4a49-8295-3aadcd9164a8">

Please see image attached.

Am I doing something wrong?

Thanks

Alan

**Comment by SchneiderInfosystems** [comment, marked as answer]
Hi Alan

That has changed while correction of https://github.com/SchneiderInfosystems/FB4D/issues/24 already in September 2019 still before version 1.0 was published. On the other hand, I can confirm that the deprecated version of `TFirestoreDatabase.InsertOrUpdateDocument `did not encode the path correctly when specifying the path, which I now consider a bug. The `class function TFirestoreDocument.GetDocFullPath` should encode the document path identical with `TFirebaseHelpers.EncodeResourceParams`.

The firestore spec would allow a space (https://firebase.google.com/docs/firestore/quotas#collections_documents_and_fields), but from my point of view a space in an ID is not a good design. How you can you easily distinguish between one, two or more spaces?

If you need the space in IDs for compatibility reasons, remove the entry `TURLEncoding.TEncodeOption.SpacesAsPlus` in _FB4.Helpers_ in line 372 in method `TFirebaseHelpers.EncodeResourceParams` in the constant `Option`. But bear in mind that this central routine is also used in other places where it could cause problems (e.g. in storage). 

I could offer this as an option, but the interface doesn't get any easier with such a change and as I said, a good design prevents spaces in IDs. 

Christoph

**Comment by JediThumper** [comment]
Hi Christoph,
Many thanks for your reply.

I understand exactly what you are saying about spaces not being a good idea and will change my code to reflect your advice.

However, I use a date time for a document name which also puts a + in the space.

Do you have a suggestion on how to deal with this? I do remember reading something about dates in document names but cant remember.

Thank you again

Alan 
![Screenshot](https://github.com/SchneiderInfosystems/FB4D/assets/71242335/c52b969d-3368-4979-b872-4647f6d35955)

**Comment by SchneiderInfosystems** [comment]
I also have a project where I use a timestamp aso document ID to sort the documents by creation time. However, I use `CreateAutoID(PUSHID)` where only the first 8 characters are used for the date/time followed by 12 random characters, which reliably prevent documents created simultaneously by several senders from having the same ID. Additionally, and to improve readability, I have added a field into this document with the creation date that is easy to read. Your approach of using date/time as the document ID will of course only work if you are sure that multiple documents are not created per minute. You can simply convert the new plus back into a space when reading the document, because the plus will not occur in date and time. This also guarantees backwards compatibility with existing documents in the same Firestore database.

I have now corrected that paths are also encoded correctly in the deprecated method ```
TFirestoreDatabase.InsertOrUpdateDocument(
  DocumentPath..)
```, because this problem has not only affected the space character, but also "<>^`{}|/\?#+%, which can all also cause problems. 
This means that all methods are encoded the path now symmetrically in the same way, which is important if you use a listener on these documents, for example.

**Reply by JediThumper** [reply]
Hi Christoph,

Thanks again for your answer and updates.

Alan
